libr/include/r_util/r_assert.h

```diff
@@ -0,0 +1,116 @@
+#ifndef R_ASSERT_H
+#define R_ASSERT_H
+
+#include "r_log.h"
+
+R_API void r_assert_log(RLogLevel level, const char *fmt, ...);
+
+#if defined (__GNUC__) && defined (__cplusplus)
+#define R_FUNCTION     ((const char*) (__PRETTY_FUNCTION__))
+#elif defined (__STDC_VERSION__) && __STDC_VERSION__ >= 199901L
+#define R_FUNCTION     ((const char*) (__func__))
+#elif defined (__GNUC__) || (defined(_MSC_VER) && (_MSC_VER > 1300))
+#define R_FUNCTION     ((const char*) (__FUNCTION__))
+#else
+#define R_FUNCTION     ((const char*) ("???"))
+#endif
+
+#define r_warn_if_reached()						\
+	do {								\
+		r_assert_log (R_LOG_WARNING, "(%s:%d):%s%s code should not be reached\n", \
+			__FILE__, __LINE__, R_FUNCTION, R_FUNCTION[0] ? ":" : ""); \
+	} while (0)
+
+#define r_warn_if_fail(expr)					\
+	do {								\
+		if (!(expr)) {						\
+			r_assert_log (R_LOG_WARNING, "WARNING (%s:%d):%s%s runtime check failed: (%s)\n", \
```

Comment:
you removed my comments on moving this into a separate function to avoid having the same string on every single object inside r2

Reply:
i already answered. That's not going to happen. The compiler is smart enough to put just one instance of the string in one place, it won't create a new string for every use of this macro.

Reply:
i know its not going to be appearing more than once per object. But we can have it in only 1 object and not in all of them, thats the improvement i was asking for. also is better for partial rebuilds when doing changes on those macros (which shouldnt happen frequently), but will make compiler life's easier

Reply:
No, it's not going to appear 1 per library. And really, do we care if there's a string of length 10 in a library? Really not necessary IMHO.

Reply:
Just for clarity: if you compile in Debug mode with no optimizations, there will be a string for each *object*, but as soon as you add any optimization (e.g. in meson you use `--buildtype debugoptimized` or any of the other non-debug modes) the compiler will merge the same strings and just use one string for each *library*.

I personally see it as expected, but given what you previously said I guess you'd like to have also debug mode as optimized as possible. So I'm going to move the string inside a function. This approach, however, creates a different function for each string we may want to support, which seems wrong to me.